src/pathfinding_igvc/src/PathFinding.cpp

```diff
@@ -0,0 +1,144 @@
+/*
+ * Created By: William Gu
+ * Created On: Jan 20, 2018
+ * Description: A path finding node which converts a given path into a Twist
+ * message to send to the robot
+ */
+
+#include <PathFinding.h>
+
+PathFinding::PathFinding(int argc, char** argv, std::string node_name) {
+    // Setup NodeHandles
+    ros::init(argc, argv, node_name);
+    ros::NodeHandle nh;
+    ros::NodeHandle private_nh("~");
+    uint32_t queue_size = 1;
+
+    /*Assume initial coordinates: may need to change later*/
+    robot_x_pos      = 0;
+    robot_y_pos      = 0;
+    robotOrientation = 0;
+
+    std::string path_subscribe_topic = "/path"; // Setup subscriber to path
+    path_subscriber                  = nh.subscribe(
+    path_subscribe_topic, queue_size, &PathFinding::pathCallBack, this);
+
+    std::string tf_subscribe_topic = "/tf"; // Setup subscriber to tf
+    tf_subscriber                  = nh.subscribe(
+    tf_subscribe_topic, queue_size, &PathFinding::tfCallBack, this);
+
+    std::string twist_publish_topic =
+    private_nh.resolveName("/cmd_vel"); // setup Publisher to twist
+    twist_publisher =
+    nh.advertise<geometry_msgs::Twist>(twist_publish_topic, queue_size);
+
+    // Get Params
+    SB_getParam(
+    private_nh, "base_frame", base_frame, (std::string) "base_link");
+    SB_getParam(
+    private_nh, "global_frame", global_frame, (std::string) "odom_combined");
+    SB_getParam(private_nh, "num_poses", num_poses, 10);
+}
+
+void PathFinding::pathCallBack(const nav_msgs::Path::ConstPtr& path_ptr) {
+    nav_msgs::Path path_msg =
+    *path_ptr; // Take required information from received message
+    geometry_msgs::Twist twist_msg = pathToTwist(
+    path_msg, robot_x_pos, robot_y_pos, robotOrientation, num_poses);
```

Comment:
What about the case where we receive a path before a tfcallback? Won't `robot_x_pos` and `robot_y_pos` still be 0?

Reply:
I was thinking of adding some type of flag variable that's set to true once we receive the first tfcallback and then checking the flag for each pathcallback so that a twist message won't be published if the flag is false. I don't know if there's a better way to do this though, since we would need to check the flag for every pathcallback.

Reply:
That seems like a pretty reasonable approach to me!